codejail/jail.py

```diff
@@ -0,0 +1,329 @@
+    """
+    default_language = languages.other
+
+    known_commands = {
+        'python': languages.python2,
+    }
+
+    def __init__(self, command, bin_path, user, lang=None):
+        self.command = command
+        self.bin_path = bin_path
+        self.user = user
+        if lang is None:
+            self.lang = self.known_commands.get(command, self.default_language)
+        else:
+            self.lang = lang
+
+    @property
+    def cmdline_start(self):
+        """
+        The command line for this program.
+        """
+        return [self.bin_path] + self.lang.argv
+
+    def __getitem__(self, item):
+        """
+        For refactoring ease, mimic the old COMMANDS[x] dict
+        """
+        if item == 'user':
+            return self.user
+        elif item == 'cmdline_start':
+            return self.cmdline_start
+        else:
+            raise KeyError(item)
+
+    def safe_exec(self, code, globals_dict, files=None, python_path=None, slug=None, extra_files=None):
+        """
+        Execute code as "exec" does, but safely.
+
+        `code` is a string of Python code.  `globals_dict` is used as the globals
+        during execution.  Modifications the code makes to `globals_dict` are
+        reflected in the dictionary on return.
+
+        `files` is a list of file paths, either files or directories.  They will be
+        copied into the temp directory used for execution.  No attempt is made to
+        determine whether the file is appropriate or safe to copy.  The caller must
+        determine which files to provide to the code.
+
+        `python_path` is a list of directory or file paths.  These names will be
+        added to `sys.path` so that modules they contain can be imported.  Only
+        directories and zip files are supported.  If the name is not provided in
+        `extra_files`, it will be copied just as if it had been listed in `files`.
+
+        `slug` is an arbitrary string, a description that's meaningful to the
+        caller, that will be used in log messages.
+
+        `extra_files` is a list of pairs, each pair is a filename and a bytestring
+        of contents to write into that file.  These files will be created in the
+        temp directory and cleaned up automatically.  No subdirectories are
+        supported in the filename.
+
+        Returns None.  Changes made by `code` are visible in `globals_dict`.  If
+        the code raises an exception, this function will raise `SafeExecException`
+        with the stderr of the sandbox process, which usually includes the original
+        exception message and traceback.
+        """
+        if self.lang.safe_exec_template is None:
+            error_template = "Code Jail {} not configured to run in safe_exec mode.  Try jail_code() instead"
+            raise JailError(error_template.format(self.command))
+        files = list(files or ())
+        extra_files = extra_files or ()
+        python_path = python_path or ()
```

Comment:
Are the python specifics appropriate here, method name is generic, we support, theoretically, other languages.